docs/development/best-practices.md

```diff
@@ -0,0 +1,115 @@
+# Best Practices
+
+## General
+
+- Prefer full function declaration, as opposed to `const func = ():void => {}`, for readability and better stack traces
+- Avoid [Enums](https://github.com/renovatebot/renovate/issues/13743), use union
+  or [immutable objects](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/pep440/range.ts#L8-L20) instead
+- Always add unit tests for full code coverage
+  - Only use `istanbul` comments for unreachable code coverage that is needed for `codecov` completion.
+  - Use descriptive `istanbul` comments
+
+```js
+// istanbul ignore next: can never happen
+```
+
+### Logging
+
+- Use logger metadata:
+
+Use
+
+```js
+logger.trace({ obj }, 'title');
+```
+
+instead of
+
+```js
+logger.trace('title:\n' + obj);
+```
+
+### Array constructor
+
+Avoid the `Array()` constructor, with or without `new`, in your TypeScript code.
+It has confusing and contradictory usage.
+
+```js
+const a = new Array(2); // [undefined, undefined]
+const b = new Array(2, 3); // [2, 3];
+```
+
+Instead, always use bracket notation to initialize arrays, or `from` to initialize an Array with a certain size.
```

Comment:
"Initialize an Array with a certain size" is a bit vague to me. Are there guidelines when you should use Bracket notatation, and when you should use `from`? Can we give a number of array items when you should use `from`?

Reply:
added a `from` example